lib/internal/repl.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+
+module.exports = {createRepl: createRepl};
+
+const Interface = require('readline').Interface;
+const REPL = require('repl');
+const path = require('path');
+
+try {
+  // hack for require.resolve("./relative") to work properly.
+  module.filename = path.resolve('repl');
+} catch (e) {
+  // path.resolve('repl') fails when the current working directory has been
+  // deleted.  Fall back to the directory name of the (absolute) executable
+  // path.  It's not really correct but what are the alternatives?
+  const dirname = path.dirname(process.execPath);
+  module.filename = path.resolve(dirname, 'repl');
+}
+
+// hack for repl require to work properly with node_modules folders
+module.paths = require('module')._nodeModulePaths(module.filename);
+
+function createRepl(env, cb) {
+  const opts = {
+    useGlobal: true,
+    ignoreUndefined: false
+  };
+
+  if (parseInt(env.NODE_NO_READLINE)) {
+    opts.terminal = false;
+  }
+  if (parseInt(env.NODE_DISABLE_COLORS)) {
+    opts.useColors = false;
+  }
+
+  opts.replMode = {
+    'strict': REPL.REPL_MODE_STRICT,
+    'sloppy': REPL.REPL_MODE_SLOPPY,
+    'magic': REPL.REPL_MODE_MAGIC
+  }[String(env.NODE_REPL_MODE).toLowerCase()];
+
+  if (opts.replMode === undefined) {
+    opts.replMode = REPL.REPL_MODE_MAGIC;
+  }
+
+  const historySize = Number(env.NODE_REPL_HISTORY_SIZE);
+  if (!isNaN(historySize)) {
+    opts.historySize = historySize;
+  }
+
+  const repl = REPL.start(opts);
+  if (env.NODE_REPL_HISTORY_PATH) {
+    return setupHistory(repl, env.NODE_REPL_HISTORY_PATH, cb);
```

Comment:
From our conversation in IRC, I think I'm going to punt on this – a setuid'd binary running a repl will be dangerous no matter what; I think the better solution is to detect those situations and refuse to run the REPL at all (or, if we can get away with it, the entire program!) Once #1536 lands we should be able to open a new PR introducing those changes.